File: controllers/autonomous_python/mavic.py
from controller import Robot, Camera, GPS, InertialUnit, Motor, LED
from typing import Tuple

from constants import (K_VERTICAL_THRUST, K_VERTICAL_OFFSET,
                       K_VERTICAL_P, K_VERTICAL_I, K_VERTICAL_D,
                       K_ROLL_P, K_PITCH_P, MAX_SIMULATION_TIME, MAX_MOTOR_VELOCITY)

class Mavic:
    def __init__(self):
        
        ## initialize Robot
        self.drone = Robot()
        self.timestep = int(self.drone.getBasicTimeStep())

        ## Get Camera
        self.camera = self.drone.getDevice("camera")
        self.camera.enable(self.timestep)

        ## Get IMU
        self.imu = self.drone.getDevice("inertial unit")
        self.imu.enable(self.timestep)

        ## Get gyro
        self.gyro = self.drone.getDevice("gyro")
        self.gyro.enable(self.timestep)
        
        ## Get GPS
        self.gps = self.drone.getDevice("gps")
        self.gps.enable(self.timestep)

        ## Get Propellers
        self.front_left_motor = self.drone.getDevice("front left propeller")
        self.front_right_motor = self.drone.getDevice("front right propeller")
        self.rear_left_motor = self.drone.getDevice("rear left propeller")
        self.rear_right_motor = self.drone.getDevice("rear right propeller")

        self.motors = [
            self.front_left_motor,
            self.front_right_motor,
            self.rear_left_motor,
            self.rear_right_motor
        ]

        for motor in self.motors:
            motor.setPosition(float('inf'))  # Set motors to velocity mode
            motor.setVelocity(1.0)  # Initial velocity for motors

    def get_imu_values(self) -> Tuple[float, float, float]:
        return self.imu.getRollPitchYaw()
    
    def get_gps_values(self) -> Tuple[float, float, float]:
        return self.gps.getValues()
    
    def get_gyro_values(self) -> Tuple[float, float, float]:
        return self.gyro.getValues()
    
    def get_time(self) -> float:
        return self.drone.getTime()
    
    def set_rotor_speed(self, speed: Tuple[float, float, float, float]) -> None:
        fl_speed, fr_speed, rl_speed, rr_speed = speed

        self.front_left_motor.setVelocity(fl_speed)
        self.front_right_motor.setVelocity(fr_speed)
        self.rear_left_motor.setVelocity(rl_speed)
        self.rear_right_motor.setVelocity(rr_speed)

File: controllers/autonomous_cpp/Mavic.cpp
#include "Mavic.hpp"

#include <iostream>
#include <tuple>
#include <limits>

#include <webots/Robot.hpp>
#include <webots/Camera.hpp>
#include <webots/GPS.hpp>
#include <webots/Gyro.hpp>
#include <webots/InertialUnit.hpp>

Mavic::Mavic() {
    // Initialize Robot
    timestep = getBasicTimeStep();

    // Get Camera
    camera = getCamera("camera");
    camera->enable(timestep);

    // Get IMU
    imu = getInertialUnit("inertial unit");
    imu->enable(timestep);

    // Get Gyro
    gyro = getGyro("gyro");
    gyro->enable(timestep);

    // Get GPS
    gps = getGPS("gps");
    gps->enable(timestep);

    // Get Propellers
    front_left_motor = getMotor("front left propeller");
    front_left_motor->setPosition(std::numeric_limits<double>::infinity());
    front_left_motor->setVelocity(1.);

    front_right_motor = getMotor("front right propeller");
    front_right_motor->setPosition(std::numeric_limits<double>::infinity());
    front_right_motor->setVelocity(1.);

    rear_left_motor = getMotor("rear left propeller");
    rear_left_motor->setPosition(std::numeric_limits<double>::infinity());
    rear_left_motor->setVelocity(1.);

    rear_right_motor = getMotor("rear right propeller");
    rear_right_motor->setPosition(std::numeric_limits<double>::infinity());
    rear_left_motor->setVelocity(1.);

}

const double* Mavic::get_imu_values() {
    return imu->getRollPitchYaw();
}

const double* Mavic::get_gps_values() {
    return gps->getValues();
}

const double* Mavic::get_gyro_values() {
    return gyro->getValues();
}

double Mavic::get_time() {
    return getTime();
}

void Mavic::set_rotor_speed(const double* rotor_speeds) {
    double fl_speed, fr_speed, rl_speed, rr_speed;
    fl_speed = rotor_speeds[0];
    fr_speed = rotor_speeds[1];
    rl_speed = rotor_speeds[2];
    rr_speed = rotor_speeds[3];

    front_left_motor->setVelocity(fl_speed);
    front_right_motor->setVelocity(fr_speed);
    rear_left_motor->setVelocity(rl_speed);
    rear_right_motor->setVelocity(rr_speed);
}